src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayer.java

```diff
@@ -221,6 +222,52 @@ public void setTotalExperience(int exp) {
         this.experienceTotal = exp;
     }
 
+    /**
+     * {@link EntityPlayer#addExperienceLevel(int)} and
+     * {@link EntityPlayer#removeExperienceLevel(int)} don't update the total
+     * experience. This recalculates it for plugins to properly make use of it.
+     *
+     * @see #onAddExperienceLevels(int, CallbackInfo)
+     */
+    private void recalculateTotalExperience() {
+        if (!this.dontRecalculateExperience) {
+            this.experienceTotal = this.xpAtLevel(this.experienceLevel) + Math.round(this.experience * this.xpBarCap());
```

Comment:
Oh, I agree, the system is badly done, but we should still try to keep things in sync as much as possible.

With regards to `this.experience`, if you leave it alone after performing your experience calculation, you run the risk of desyncing again further down the road. The chances of this getting to be a large issue isn't high, but because you have a little bit extra on the XP bar, once each XP point represents less than this extra you've got, (which can happen if you just continue to add experience by points alone using the standard Minecraft method) you'll start to level up early again. I've not had time to test this yet, but it certainly is theoretically possible, which is why I think we _should_ be updating it once the `totalExperience` is synced up, if only for this reason alone.

Total experience isn't affected by not resetting `this.experience`, but the leveling mechanics are. 

> Vanilla Minecraft would say round down.

Then that's what we should be doing, realistically. Don't get me wrong, I understand your point and reasoning, but for a quick fix in the case of adding no levels, that is, the experience is already OK for the current level, is that we should just not recalculate the experience. It doesn't completely remove the issue that changing levels might be out by an XP point, sure, but it does remove the obviousness if someone doesn't actually change levels.

Maybe it's worth recalculating `this.experience` when completing `addExperience` to reduce the inaccuracy?